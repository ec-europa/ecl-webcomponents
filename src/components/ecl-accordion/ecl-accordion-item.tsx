import { Component, Prop, Element, h, Event, EventEmitter } from '@stencil/core';

@Component({
  tag: 'ecl-accordion-item',
  styleUrl: 'build/styles/ecl-accordion-item.css',
  shadow: false,
  scoped: false,
  assetsDirs: ['build'],
})

export class EclAccordionItem {
  @Element() el: HTMLElement;
  @Prop() styleClass: string;
  @Prop() label: string;
  @Prop() itemId: string;
  @Prop({reflect:true}) expanded: boolean;
  @Prop() theme: string;
  @Event() toggleItem: EventEmitter<string>;

  handleClick = () => {
    this.toggleItem.emit(this.itemId);
  };
  
  getClass(): string {
    return [
      `ecl-accordion__item`,
      `sc-ecl-accordion-${this.theme}`,
      this.styleClass
    ].join(' ');
  }

  componentWillLoad() {
    this.theme = document.documentElement.getAttribute('data-ecl-theme') ?? (this.theme || 'ec');
  }

  render() {
    const titleId = `${this.itemId}-title`;
    const contentId = `${this.itemId}-content`;
    return (
      <div
        class={this.getClass()}
      >
        <div
          class={`ecl-accordion__title sc-ecl-accordion-${this.theme}`}
          id={titleId}
        >
          <button
            type="button"
            data-ecl-accordion-toggle
            onClick={this.handleClick}
            aria-controls={contentId}
            aria-expanded={this.expanded ? 'true' : 'false'}
            class={`ecl-accordion__toggle sc-ecl-accordion-${this.theme}`}
          >
            <span class={`ecl-accordion__toggle-flex sc-ecl-accordion-${this.theme}`}>
              <span class={`ecl-accordion__toggle-title sc-ecl-accordion-${this.theme}`}>
                {this.label}
              </span>
              <span class={`ecl-accordion__toggle-indicator sc-ecl-accordion-${this.theme}`}>
                <ecl-icon
                  icon={this.expanded ? "minus" : 'plus'}
                  theme={this.theme}
                  size="s"
                  style-class={`ecl-accordion__toggle-icon sc-ecl-accordion-${this.theme}`}
                  data-ecl-accordion-icon
                ></ecl-icon>
              </span>
            </span>
          </button>
        </div>
        <div 
          class={`ecl-accordion__content sc-ecl-accordion-${this.theme}`}
          role="region"
          id={contentId}
          aria-labelledby={titleId}
          hidden={this.expanded !== true}
        >
          <slot></slot>
        </div>
      </div>
    )
  }
}