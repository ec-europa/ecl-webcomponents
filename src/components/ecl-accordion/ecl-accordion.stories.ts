export default {
  title: 'Components/accordion',
};

const Template = (args) => 
  `<ecl-accordion 
    ecl-script
    color-mode="${args.color_mode}"
  >
    <ecl-accordion-item
      item-id="ecl-accordion-item-id-1"
      label="Delivery of last pending proposals, a common Destiny of unity, the hour of European Democracy"
    >
      The College of Commissioners held today the first weekly meeting of 2019 which was devoted to discussing the
      challenges of this new year. Commissioners used the opportunity to take stock and discuss the year ahead,
      including the European elections in May and other important milestones ahead.
    </ecl-accordion-item>

    <ecl-accordion-item
      item-id="ecl-accordion-item-id-2"
      label="Energy union and climate"
      expanded
    >
      A balanced and progressive trade policy to harness globalisation
    </ecl-accordion-item>

    <ecl-accordion-item
      item-id="ecl-accordion-item-id-3"
      label="Delivery of last pending proposals, a common Destiny of unity, the hour of European Democracy"
    >
      In the modern global economy trade is essential for growth, <ecl-link path="https://ec.europa.eu/">jobs</ecl-link> and competiveness, and the EU is committed to
      maintaining an open and rules-based trading system. With the rising threat of protectionism and weakened
      commitment of large players to global trade governance, the EU must take the lead.
    </ecl-accordion-item>
  </ecl-accordion>`;

export const Accordion = Template.bind({});
Accordion.storyName = 'default';
