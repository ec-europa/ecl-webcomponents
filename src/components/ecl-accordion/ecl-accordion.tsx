import { Component, Prop, h, Element } from '@stencil/core';
import getAssetPath from "../../utils/assetPath";
declare const ECL: any;

@Component({
  tag: 'ecl-accordion',
  styleUrls: {
    ec: './build/styles/ecl-accordion-ec.css',
    eu: './build/styles/ecl-accordion-eu.css',
  },
  shadow: false,
  scoped: true,
  assetsDirs: ['build'],
})

export class EclAccordion {
  @Element() el: HTMLElement;
  @Prop() styleClass: string = '';
  @Prop() eclScript: boolean = false;
  @Prop() theme: string = 'ec';
  @Prop() clickedItem: HTMLElement;

  getClass(): string {
    return [
      `ecl-accordion`,
      this.styleClass
    ].join(' ');
  }

  componentWillLoad() {
    if (this.eclScript) {
      const src = getAssetPath('./build/scripts/ecl-accordion-vanilla.js');
      if (document.querySelector(`script[src="${src}"]`)) {
        document.querySelector(`script[src="${src}"]`).remove();
      }
      const script = document.createElement('script');
      script.src = src;
      script.onload = () => {
        const accordion = new ECL.Accordion(this.el);
        accordion.init();
      };
      document.body.appendChild(script);
    }
  }

  componentDidRender() {
    this.el.querySelectorAll('.ecl-accordion__item button').forEach((item) => {
      item.addEventListener('click', (e) => {
        const target = e.currentTarget as HTMLButtonElement;
        this.clickedItem = target.parentElement.parentElement;
      });
    });
  }

  render() {
    return (
      <div
        class={this.getClass()}
      >
        <slot></slot> 
      </div>
    )
  }
}
