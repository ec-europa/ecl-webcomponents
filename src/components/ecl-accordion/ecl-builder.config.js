const path = require('path');
const pkg = require('./package.json');

const nodeModules = path.resolve(__dirname, './node_modules');

const outputFolder = './build';

// SCSS includePaths
const includePaths = [nodeModules];

const banner = `${pkg.name} - ${pkg.version} Built on ${new Date().toISOString()}`;

module.exports = {
  copy: [
    {
      from: [path.resolve(nodeModules, '@ecl/resources-icons/dist/sprites/icons.svg')],
      to: path.resolve(outputFolder, 'images/ec'),
      options: { up: true },
    },
    {
      from: [path.resolve(nodeModules, '@ecl/resources-icons/dist/sprites/icons.svg')],
      to: path.resolve(outputFolder, 'images/eu'),
      options: { up: true },
    },
  ],
  scripts: [
    {
      entry: path.resolve(nodeModules, '@ecl/accordion/accordion.js'),
      dest: path.resolve(outputFolder, 'scripts/ecl-accordion-vanilla.js'),
      options: {
        banner,
        format: 'iife',
        moduleName: 'ECL',
        includePaths,
        sourceMap: false,
      },
    },
  ],
  styles: [
    {
      entry: path.resolve(__dirname, 'src/styles/ecl-accordion-ec.scss'),
      dest: path.resolve(outputFolder, 'styles/ecl-accordion-ec.css'),
      options: {
        banner,
        includePaths,
        sourceMap: false,
        minify: true,
      },
    },
    {
      entry: path.resolve(__dirname, 'src/styles/ecl-accordion-item.scss'),
      dest: path.resolve(outputFolder, 'styles/ecl-accordion-item.css'),
      options: {
        banner,
        includePaths,
        sourceMap: false,
        minify: true,
      },
    },
    {
      entry: path.resolve(__dirname, 'src/styles/ecl-accordion-eu.scss'),
      dest: path.resolve(outputFolder, 'styles/ecl-accordion-eu.css'),
      options: {
        banner,
        includePaths,
        sourceMap: false,
        minify: true,
      },
    },
  ],
};
