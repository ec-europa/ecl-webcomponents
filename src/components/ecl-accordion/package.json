{
  "name": "@ecl/web-components-accordion",
  "author": "European Commission",
  "license": "EUPL-1.2",
  "version": "0.12.0",
  "description": "ECL Button",
  "main": "ecl-accordion.tsx",
  "private": true,
  "scripts": {
    "build:copy": "cross-env ecl-builder copy",
    "build:styles": "cross-env ecl-builder styles",
    "build:scripts": "cross-env ecl-builder scripts",
    "build": "run-p clean build:*",
    "clean": "rimraf build",
    "prepublish": "npm run build"
  },
  "devDependencies": {
    "@ecl/builder": "3.13.0",
    "@ecl/resources-ec-icons": "3.13.0",
    "@ecl/resources-eu-icons": "3.13.0",
    "@ecl/specs-component-accordion": "3.13.0",
    "@ecl/theme-ec": "3.13.0",
    "@ecl/theme-eu": "3.13.0",
    "@ecl/vanilla-component-accordion": "3.13.0",
    "@ecl/vanilla-component-button": "3.13.0",
    "@ecl/vanilla-component-icon": "3.13.0",
    "@ecl/vanilla-layout-grid": "3.13.0",
    "cross-env": "7.0.3"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/ec-europa/europa-component-library.git"
  },
  "bugs": {
    "url": "https://github.com/ec-europa/europa-component-library/issues"
  },
  "homepage": "https://github.com/ec-europa/europa-component-library",
  "keywords": [
    "ecl",
    "europa-component-library",
    "design-system",
    "web-components"
  ]
}
