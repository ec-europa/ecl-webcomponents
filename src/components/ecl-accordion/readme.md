# ecl-accordion-item



<!-- Auto Generated Below -->


## Properties

| Property     | Attribute     | Description | Type      | Default     |
| ------------ | ------------- | ----------- | --------- | ----------- |
| `expanded`   | `expanded`    |             | `boolean` | `undefined` |
| `itemId`     | `item-id`     |             | `string`  | `undefined` |
| `label`      | `label`       |             | `string`  | `undefined` |
| `styleClass` | `style-class` |             | `string`  | `undefined` |
| `theme`      | `theme`       |             | `string`  | `undefined` |


## Events

| Event        | Description | Type                  |
| ------------ | ----------- | --------------------- |
| `toggleItem` |             | `CustomEvent<string>` |


## Dependencies

### Depends on

- [ecl-icon](../ecl-icon)

### Graph
```mermaid
graph TD;
  ecl-accordion-item --> ecl-icon
  style ecl-accordion-item fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
