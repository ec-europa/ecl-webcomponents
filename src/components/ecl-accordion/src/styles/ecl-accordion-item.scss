:host {
  display: block;
}

.ecl-accordion__item.sc-ecl-accordion-ec {
  border-bottom-width: 1px !important;
}

.ecl-accordion__item.sc-ecl-accordion-eu {
  border-bottom-width: 2px !important;
}

.ecl-accordion__item.sc-ecl-accordion-eu:not(.is-first)::before {
  display: none;
}

.ecl-accordion__item.is-last {
  border-bottom: none !important;
}

.ecl-accordion__item.sc-ecl-accordion-eu:not(.is-first) .ecl-accordion__toggle.sc-ecl-accordion-eu {
  border-radius: 0 !important;
}

.ecl-accordion__item.sc-ecl-accordion-eu.is-last .ecl-accordion__toggle.sc-ecl-accordion-eu {
  border-end-end-radius: var(--s-xs) !important;
  border-end-start-radius: var(--s-xs) !important;
}

.ecl-accordion__item.sc-ecl-accordion-eu.is-first .ecl-accordion__toggle.sc-ecl-accordion-eu {
  border-end-end-radius: 0 !important;
  border-end-start-radius: 0 !important;
}
