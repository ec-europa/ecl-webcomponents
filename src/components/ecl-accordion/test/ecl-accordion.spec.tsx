import { newSpecPage } from '@stencil/core/testing';
import { EclAccordion } from '../ecl-accordion';
import { EclAccordionItem } from '../ecl-accordion-item';

describe('ecl-accordion', () => {
  it('renders', async () => {
    const page = await newSpecPage({
      components: [EclAccordion, EclAccordionItem],
      html: `<ecl-accordion theme="ec">
              <ecl-accordion-item
                id="ecl-accordion-item-id-1"
                label="Delivery of last pending proposals, a common Destiny of unity, the hour of European Democracy"
                icon-path="/icons.svg"
              >
                Accordion item label
              </ecl-accordion-item>
              <ecl-accordion-item
                id="ecl-accordion-item-id-2"
                label="Energy union and climate"
                icon-path="/icons.svg"
                expanded
              >
                Accordion item label
              </ecl-accordion-item>
              <ecl-accordion-item
                id="ecl-accordion-item-id-3"
                label="Delivery of last pending proposals, a common Destiny of unity, the hour of European Democracy"
              >
                Accordion item label
              </ecl-accordion-item>
             </ecl-accordion>`,
    });

    expect(page.root).toMatchSnapshot();
  });
});
