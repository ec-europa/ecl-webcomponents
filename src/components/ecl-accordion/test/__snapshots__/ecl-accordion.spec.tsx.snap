// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ecl-accordion renders 1`] = `
<ecl-accordion theme="ec">
  <!---->
  <div class="ecl-accordion">
    <ecl-accordion-item icon-path="/icons.svg" item-id="ecl-accordion-item-id-1" label="Delivery of last pending proposals, a common Destiny of unity, the hour of European Democracy" label-collapsed="Open" label-expanded="Close">
      <!---->
      <div class="ecl-accordion__item is-first is-last sc-ecl-accordion-ec">
        <div class="ecl-accordion__title sc-ecl-accordion-ec" id="ecl-accordion-item-id-1-title">
          <button aria-controls="ecl-accordion-item-id-1-content" aria-expanded="false" class="ecl-accordion__toggle sc-ecl-accordion-ec" data-ecl-accordion-toggle type="button">
            <span class="ecl-accordion__toggle-flex sc-ecl-accordion-ec">
              <span class="ecl-accordion__toggle-title sc-ecl-accordion-ec">
                Delivery of last pending proposals, a common Destiny of unity, the hour of European Democracy
              </span>
              <span class="ecl-accordion__toggle-indicator sc-ecl-accordion-ec">
                <ecl-icon data-ecl-accordion-icon style-class="ecl-accordion__toggle-icon sc-ecl-accordion-ec">
                  <svg class="ecl-accordion__toggle-icon ecl-icon ecl-icon--s sc-ecl-accordion-ec">
                    <use xlink:href="/build/images/ec/icons.svg#plus"></use>
                  </svg>
                </ecl-icon>
              </span>
            </span>
          </button>
        </div>
        <div aria-labelledby="ecl-accordion-item-id-1-title" class="ecl-accordion__content sc-ecl-accordion-ec" hidden="" id="ecl-accordion-item-id-1-content" role="region">
          Accordion item label
        </div>
      </div>
    </ecl-accordion-item>
    <ecl-accordion-item expanded="" icon-path="/icons.svg" item-id="ecl-accordion-item-id-2" label="Energy union and climate" label-collapsed="Open" label-expanded="Close">
      <!---->
      <div class="ecl-accordion__item is-first is-last sc-ecl-accordion-ec">
        <div class="ecl-accordion__title sc-ecl-accordion-ec" id="ecl-accordion-item-id-2-title">
          <button aria-controls="ecl-accordion-item-id-2-content" aria-expanded="true" class="ecl-accordion__toggle sc-ecl-accordion-ec" data-ecl-accordion-toggle type="button">
            <span class="ecl-accordion__toggle-flex sc-ecl-accordion-ec">
              <span class="ecl-accordion__toggle-title sc-ecl-accordion-ec">
                Energy union and climate
              </span>
              <span class="ecl-accordion__toggle-indicator sc-ecl-accordion-ec">
                <ecl-icon data-ecl-accordion-icon style-class="ecl-accordion__toggle-icon sc-ecl-accordion-ec">
                  <svg class="ecl-accordion__toggle-icon ecl-icon ecl-icon--s sc-ecl-accordion-ec">
                    <use xlink:href="/build/images/ec/icons.svg#minus"></use>
                  </svg>
                </ecl-icon>
              </span>
            </span>
          </button>
        </div>
        <div aria-labelledby="ecl-accordion-item-id-2-title" class="ecl-accordion__content sc-ecl-accordion-ec" id="ecl-accordion-item-id-2-content" role="region">
          Accordion item label
        </div>
      </div>
    </ecl-accordion-item>
    <ecl-accordion-item item-id="ecl-accordion-item-id-3" label="Delivery of last pending proposals, a common Destiny of unity, the hour of European Democracy" label-collapsed="Open" label-expanded="Close">
      <!---->
      <div class="ecl-accordion__item is-first is-last sc-ecl-accordion-ec">
        <div class="ecl-accordion__title sc-ecl-accordion-ec" id="ecl-accordion-item-id-3-title">
          <button aria-controls="ecl-accordion-item-id-3-content" aria-expanded="false" class="ecl-accordion__toggle sc-ecl-accordion-ec" data-ecl-accordion-toggle type="button">
            <span class="ecl-accordion__toggle-flex sc-ecl-accordion-ec">
              <span class="ecl-accordion__toggle-title sc-ecl-accordion-ec">
                Delivery of last pending proposals, a common Destiny of unity, the hour of European Democracy
              </span>
              <span class="ecl-accordion__toggle-indicator sc-ecl-accordion-ec">
                <ecl-icon data-ecl-accordion-icon style-class="ecl-accordion__toggle-icon sc-ecl-accordion-ec">
                  <svg class="ecl-accordion__toggle-icon ecl-icon ecl-icon--s sc-ecl-accordion-ec">
                    <use xlink:href="/build/images/ec/icons.svg#plus"></use>
                  </svg>
                </ecl-icon>
              </span>
            </span>
          </button>
        </div>
        <div aria-labelledby="ecl-accordion-item-id-3-title" class="ecl-accordion__content sc-ecl-accordion-ec" hidden="" id="ecl-accordion-item-id-3-content" role="region">
          Accordion item label
        </div>
      </div>
    </ecl-accordion-item>
  </div>
</ecl-accordion>
`;
